dbms/src/Storages/DeltaMerge/Segment.cpp

```diff
@@ -3243,32 +3245,38 @@ BlockInputStreamPtr Segment::getBitmapFilterInputStream(
         start_ts,
         expected_block_size,
         enable_handle_clean_read,
-        ReadTag::Query,
+        read_tag,
         is_fast_scan,
         enable_del_clean_read,
         /* read_packs */ {},
-        /* need_row_id */ false,
+        /* need_row_id */ NeedRowID,
         /* bitmap_filter */ bitmap_filter);
 
     auto columns_to_read_ptr = std::make_shared<ColumnDefines>(columns_to_read);
-    SkippableBlockInputStreamPtr delta_stream = std::make_shared<DeltaValueInputStream>(
+
+    auto memtable = segment_snap->delta->getMemTableSetSnapshot();
+    auto persisted_files = segment_snap->delta->getPersistedFileSetSnapshot();
+    SkippableBlockInputStreamPtr mem_table_stream = std::make_shared<ColumnFileSetInputStream>(
         dm_context,
-        segment_snap->delta,
+        memtable,
         columns_to_read_ptr,
         this->rowkey_range,
-        ReadTag::Query);
+        read_tag);
+    SkippableBlockInputStreamPtr persisted_files_stream = std::make_shared<ColumnFileSetInputStream>(
+        dm_context,
+        persisted_files,
+        columns_to_read_ptr,
+        this->rowkey_range,
+        read_tag);
 
-    return std::make_shared<BitmapFilterBlockInputStream>(
-        columns_to_read,
-        stable_stream,
-        delta_stream,
-        segment_snap->stable->getDMFilesRows(),
-        bitmap_filter,
-        dm_context.tracing_id);
+    auto stream = std::dynamic_pointer_cast<ConcatSkippableBlockInputStream<NeedRowID>>(stable_stream);
+    stream->appendChild(persisted_files_stream, persisted_files->getRows());
+    stream->appendChild(mem_table_stream, memtable->getRows());
```

Comment:
`persisted_files->getRows()` and `memtable->getRows()` may not be the actual rows when there are ColumnFileBig but only a part of the key-range of the DMFile is covered for this segment.
Does it affect the correctness?

Reply:
No, when building the bitmap, all physical rows will be included.